packages/core/src/amazonq/messages/messagePublisher.ts

```diff
@@ -12,3 +13,45 @@ export class MessagePublisher<T> {
         this.eventEmitter.fire(event)
     }
 }
+
+/**
+ * Same as {@link MessagePublisher}, but will wait until the UI indicates it
+ * is ready to recieve messages, before the message is published.
+ *
+ * This solves a problem when running a right click menu option like
+ * "Send To Prompt" BUT chat is not opened yet, it would result in the prompt failing to
+ * be recieved by chat.
+ */
+export class UiReadyMessagePublisher<T> extends MessagePublisher<T> {
+    constructor(
+        eventEmitter: EventEmitter<T>,
+        private readonly isChatUiReady: () => boolean
+    ) {
+        super(eventEmitter)
+    }
+    public override publish(event: T): void {
+        // immediately send if Chat UI is ready
+        if (this.isChatUiReady()) {
+            super.publish(event)
+            return
+        }
+        // Otherwise wait, since it is probably still loading
+        void waitUntil(
```

Comment:
Did you experiment with trying to use a buffer that keeps filling up until the chat ui is ready? Then you can just send an event when the chat UI is ready that publishes everthing in the buffer.

I.e

```typescript
if (this.isChatUIReady()) {
    super.publish(event)
    return
}
buffer.push(event)
```
and then you'd do something like this
```typescript
onChatReady(() => {
    for (const event of this.buffer) {
        super.publish(event)
    }
})
```

Reply:
great point, updated